module/zfs/vdev_initialize.c

```diff
@@ -628,6 +628,14 @@ vdev_initialize_thread(void *arg)
 
 		spa_config_exit(spa, SCL_CONFIG, FTAG);
 		error = vdev_initialize_ranges(vd, deadbeef);
+
+		/*
+		 * It's important to wait for any initializing writes to
+		 * by synced before allowing allocations to the metaslab.
+		 * Otherwise we may overwrite newly allocated blocks.
+		 */
+		txg_wait_synced(spa->spa_dsl_pool, 0);
```

Comment:
Nice find! I think it might be better to move this into `vdev_initialize_ms_unmark` when `ms_initializing` == 0.

Reply:
We would have to do this just prior to decrementing `ms_initializing` to 0 sincw we can't hold the `ms_lock` and do the `txg_wait_synced`.

Reply:
@grwilson I've updated the PR with this proposed change.  My original inclination was to err on the side of keeping this simple and skipping this optimization, but I'm happy with the updated version is that's preferred.